File: FakturusTrack.Android/app/src/main/java/com/fakturus/track/features/legal/ConsentManager.kt
package com.fakturus.track.features.legal

import android.content.Context
import android.util.Log
import com.fakturus.track.services.api.APIClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable

class ConsentManager(private val context: Context) {
    private val prefs = context.getSharedPreferences("consent", Context.MODE_PRIVATE)

    private val _hasRequiredConsents = MutableStateFlow(false)
    val hasRequiredConsents: StateFlow<Boolean> = _hasRequiredConsents.asStateFlow()

    private var apiClient: APIClient? = null

    fun configure(apiClient: APIClient?) {
        this.apiClient = apiClient
    }

    fun checkConsent() {
        val termsVersion = prefs.getInt("terms_version", 0)
        val privacyAcknowledged = prefs.getBoolean("privacy_acknowledged", false)
        _hasRequiredConsents.value = termsVersion > 0 && privacyAcknowledged
    }

    fun recordConsent(termsVersion: Int) {
        prefs.edit()
            .putInt("terms_version", termsVersion)
            .putBoolean("terms_synced", false)
            .putBoolean("privacy_acknowledged", true)
            .apply()
        _hasRequiredConsents.value = true
    }

    suspend fun syncPendingConsent() {
        val synced = prefs.getBoolean("terms_synced", false)
        if (synced) return
        val version = prefs.getInt("terms_version", 0)
        if (version > 0) {
            syncConsentToBackend(version)
        }
    }

    suspend fun checkForVersionUpdates() {
        val client = apiClient ?: return
        try {
            val response: LegalVersionsResponse = client.get("/api/legal/versions")
            val currentTermsVersion = prefs.getInt("terms_version", 0)

            for (doc in response.documents) {
                if (doc.type == "terms_of_service" &&
                    doc.requiresReConsent &&
                    doc.version > currentTermsVersion
                ) {
                    _hasRequiredConsents.value = false
                    return
                }
            }
        } catch (e: Exception) {
            Log.w("ConsentManager", "Version check failed", e)
        }
    }

    fun clearConsent() {
        prefs.edit().clear().apply()
        _hasRequiredConsents.value = false
    }

    private suspend fun syncConsentToBackend(termsVersion: Int) {
        val client = apiClient ?: return
        try {
            val request = ConsentSubmitRequest(
                consents = listOf(
                    ConsentSubmitItem(
                        documentType = "terms_of_service",
                        documentVersion = termsVersion,
                        consentGiven = true
                    )
                )
            )
            client.post<ConsentStatusResponse, ConsentSubmitRequest>("/api/legal/consent", request)
            prefs.edit().putBoolean("terms_synced", true).apply()
            Log.i("ConsentManager", "Consent synced to backend")
        } catch (e: Exception) {
            Log.w("ConsentManager", "Backend sync failed, will retry", e)
        }
    }
}

// API Models
@Serializable
data class LegalVersionsResponse(val documents: List<LegalDocumentInfo>)

@Serializable
data class LegalDocumentInfo(
    val type: String,
    val version: Int,
    val effectiveDate: String,
    val url: String,
    val requiresConsent: Boolean,
    val requiresReConsent: Boolean
)

@Serializable
data class ConsentSubmitRequest(val consents: List<ConsentSubmitItem>)

@Serializable
data class ConsentSubmitItem(
    val documentType: String,
    val documentVersion: Int,
    val consentGiven: Boolean
)

@Serializable
data class ConsentStatusResponse(
    val consents: List<ConsentRecord>,
    val allRequiredConsentsGiven: Boolean,
    val pendingConsents: List<String>
)

@Serializable
data class ConsentRecord(
    val documentType: String,
    val documentVersion: Int,
    val consentGiven: Boolean,
    val consentTimestamp: String
)

File: FakturusTrack.iOS/FakturusTrack/Services/Legal/ConsentManager.swift
import Foundation
import Observation

@Observable @MainActor
final class ConsentManager {
    private(set) var hasRequiredConsents = false
    private(set) var isChecking = false

    private static let termsVersionKey = "consent_terms_version"
    private static let termsSyncedKey = "consent_terms_synced"
    private static let privacyAcknowledgedKey = "consent_privacy_acknowledged"

    private var apiClient: APIClient?

    func configure(apiClient: APIClient?) {
        self.apiClient = apiClient
    }

    /// Check if all required consents are given
    func checkConsent() {
        let termsVersion = UserDefaults.standard.integer(forKey: Self.termsVersionKey)
        let privacyAcknowledged = UserDefaults.standard.bool(forKey: Self.privacyAcknowledgedKey)
        hasRequiredConsents = termsVersion > 0 && privacyAcknowledged
    }

    /// Record consent after user agrees
    func recordConsent(termsVersion: Int) {
        UserDefaults.standard.set(termsVersion, forKey: Self.termsVersionKey)
        UserDefaults.standard.set(false, forKey: Self.termsSyncedKey)
        UserDefaults.standard.set(true, forKey: Self.privacyAcknowledgedKey)
        hasRequiredConsents = true

        // Sync to backend
        Task { await syncConsentToBackend(termsVersion: termsVersion) }
    }

    /// Try to sync pending consent to backend
    func syncPendingConsent() async {
        let synced = UserDefaults.standard.bool(forKey: Self.termsSyncedKey)
        if synced { return }
        let version = UserDefaults.standard.integer(forKey: Self.termsVersionKey)
        if version > 0 {
            await syncConsentToBackend(termsVersion: version)
        }
    }

    /// Check for version updates from backend
    func checkForVersionUpdates() async {
        isChecking = true
        defer { isChecking = false }

        guard let client = apiClient else { return }

        do {
            let response: LegalVersionsAPIResponse = try await client.get("/api/legal/versions")
            let currentTermsVersion = UserDefaults.standard.integer(forKey: Self.termsVersionKey)

            for doc in response.documents {
                if doc.type == "terms_of_service" && doc.requiresReConsent && doc.version > currentTermsVersion {
                    hasRequiredConsents = false
                    return
                }
            }
        } catch {
            print("[ConsentManager] Version check failed: \(error)")
        }
    }

    /// Clear all consent data (on logout or account deletion)
    func clearConsent() {
        UserDefaults.standard.removeObject(forKey: Self.termsVersionKey)
        UserDefaults.standard.removeObject(forKey: Self.termsSyncedKey)
        UserDefaults.standard.removeObject(forKey: Self.privacyAcknowledgedKey)
        hasRequiredConsents = false
    }

    // MARK: - Private

    private func syncConsentToBackend(termsVersion: Int) async {
        guard let client = apiClient else { return }

        do {
            let request = ConsentSubmitRequest(consents: [
                ConsentSubmitItem(documentType: "terms_of_service", documentVersion: termsVersion, consentGiven: true)
            ])
            let _: ConsentStatusResponse = try await client.post("/api/legal/consent", body: request)
            UserDefaults.standard.set(true, forKey: Self.termsSyncedKey)
            print("[ConsentManager] Consent synced to backend")
        } catch {
            print("[ConsentManager] Backend sync failed, will retry: \(error)")
        }
    }
}

// MARK: - API Models

private struct LegalVersionsAPIResponse: Decodable {
    let documents: [LegalDocumentAPIInfo]
}

private struct LegalDocumentAPIInfo: Decodable {
    let type: String
    let version: Int
    let effectiveDate: String
    let url: String
    let requiresConsent: Bool
    let requiresReConsent: Bool
}

private struct ConsentSubmitRequest: Encodable {
    let consents: [ConsentSubmitItem]
}

private struct ConsentSubmitItem: Encodable {
    let documentType: String
    let documentVersion: Int
    let consentGiven: Bool
}

private struct ConsentStatusResponse: Decodable {
    let consents: [ConsentRecord]
    let allRequiredConsentsGiven: Bool
    let pendingConsents: [String]
}

private struct ConsentRecord: Decodable {
    let documentType: String
    let documentVersion: Int
    let consentGiven: Bool
    let consentTimestamp: String
}
